service/controller/v18/resource/ipam/create.go

```diff
@@ -3,18 +3,26 @@ package ipam
 import (
 	"context"
 	"fmt"
+	"math/bits"
+	"math/rand"
 	"net"
 	"reflect"
+	"time"
 
+	"github.com/giantswarm/apiextensions/pkg/apis/provider/v1alpha1"
 	"github.com/giantswarm/apiextensions/pkg/clientset/versioned"
+	"github.com/giantswarm/aws-operator/service/controller/v18/controllercontext"
+	"github.com/giantswarm/aws-operator/service/controller/v18/key"
 	"github.com/giantswarm/ipam"
 	"github.com/giantswarm/microerror"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-
-	"github.com/giantswarm/aws-operator/service/controller/v18/controllercontext"
-	"github.com/giantswarm/aws-operator/service/controller/v18/key"
 )
 
+func init() {
+	// Seed RNG for AZ shuffling.
+	rand.Seed(time.Now().UnixNano())
+}
```

Comment:
First of all, this is kind of legacy from C. There it has always worked like this (for several decades).

I originally left this out as I didn't see it as important, but Pawel noted that for e2e tests it would be good to actually get random results from the start.

When there's anything that **relies** on predictable random sequence (I'd really question why?), that code should instantiate `Rand` type instead of relying on hidden global, but here we rather want to get just random results and using global is perfectly fine. Therefore I'd rather not clutter the code by dragging unnecessary type along.

Reply:
I'm fine with that. I even prefer that than having ton of dependencies for no reason. I'd just like to have a single place where we initialize global stuff so we can check single file and see "ok that was already initialized". I think `init()` in `main.go` is a good place for that. Alternatively we can introduce `init.go` file. We could describe that in `fmt` as well. I'm happy to do so.

Reply:
With `init.go` do you mean having that naming pattern in each package? When doing these things globally in a single place I find that bad because it separates the initialization from the location of code where it's related. We would then need to comment them with pointers to code that depends on it and I can imagine these comments to get out-of-sync when refactoring code.

Reply:
I meant top level, i.e. `./aws-operator/init.go`.

Reply:
I think this could get out-of-sync with the code it's related to and needed for.

Reply:
Does that matter? We could do initializing rng in every single service and operator. That's just global thing.

Reply:
No, it doesn't matter, but then I don't see any reason why it must be moved out from here either. If thing can be initialized multiple times, it doesn't need to be in centralized place and if it cannot be initialized multiple times, then it most certainly must not be a global so this centralized `init.go` wouldn't apply to it. That's why I can't see the point in it :slightly_smiling_face:

Reply:
My point is I'd like to know how this is initialized and if it is w/o grepping the repo.
Like in this case. You could look into `init.go` in root dir and check if `rand.Seed` was already called. If it was - fine nothing to do. Otherwise you just put it here and forget about it.

But I also see your point.

I recalled another common pattern which is creating a package aliasing stdlib package. so we could just create `pkg/rand` package and do `rand.Seed` in `pkg/rand/init.go`.

Reply:
I am against all global initialization because you never know where to look for what. Everything should simply be injected via configuration. Explicitly. That is as simple as it gets and we do not have to think about it. The reason here is simplicity and consistency.

Reply:
So how I see this now: Leave it as it is now. `Rand` itself uses global, yes, but that's extremely old and known thing. Unless I'm totally wrong, it originates to C and AFAIK it behaves the same in Python, Ruby, Java, you name it. So I take that it's fine to have as-is.

Reply:
I leave it to you to decide. This LGTM from my side.